Sources/MongoSwift/ConnectionString.swift

```diff
@@ -13,199 +13,241 @@ internal class ConnectionString {
-                throw MongoError.InvalidArgumentError(message: "Failed to set directConnection to \(dc)")
-            }
-        } else if !self.hasOption("directConnection") {
-            guard mongoc_uri_set_option_as_bool(self._uri, MONGOC_URI_DIRECTCONNECTION, false) else {
-                throw MongoError.InvalidArgumentError(message: "Failed to set directConnection to false")
+    private func setBoolOption(_ name: String, to value: Bool) throws {
+        guard mongoc_uri_set_option_as_bool(self._uri, name, value) else {
+            throw self.failedToSet(name, to: value)
+        }
+    }
+
+    private func setUTF8Option(_ name: String, to value: String, redactInErrorMsg: Bool = false) throws {
+        guard mongoc_uri_set_option_as_utf8(self._uri, name, value) else {
+            throw self.failedToSet(name, to: value, redact: redactInErrorMsg)
+        }
+    }
+
+    private func setInt32Option(_ name: String, to value: Int32) throws {
+        guard mongoc_uri_set_option_as_int32(self._uri, name, value) else {
+            throw self.failedToSet(name, to: value)
+        }
+    }
+
+    /// Constructs a standardized error message about failing to set an option to a specified value. If redact=true the
+    /// value will be omitted from the message.
+    private func failedToSet(
+        _ option: String,
+        to value: CustomStringConvertible,
+        redact: Bool = false
+    ) -> MongoError.InvalidArgumentError {
+        let msg = redact ? "Failed to set \(option)" : "Failed to set \(option) to \(value)"
+        return MongoError.InvalidArgumentError(message: msg)
+    }
+
+    /// Sets and validates TLS-related on the underlying `mongoc_uri_t`.
+    private func applyAndValidateTLSOptions(_ options: MongoClientOptions?) throws {
+        if let tls = options?.tls {
+            try self.setBoolOption(MONGOC_URI_TLS, to: tls)
+        }
+
+        if let invalidCerts = options?.tlsAllowInvalidCertificates {
+            try self.setBoolOption(MONGOC_URI_TLSALLOWINVALIDCERTIFICATES, to: invalidCerts)
+        }
+
+        if let invalidHostnames = options?.tlsAllowInvalidHostnames {
+            try self.setBoolOption(MONGOC_URI_TLSALLOWINVALIDHOSTNAMES, to: invalidHostnames)
+        }
+
+        if let caFile = options?.tlsCAFile?.absoluteString {
+            try self.setUTF8Option(MONGOC_URI_TLSCAFILE, to: caFile)
+        }
+
+        if let certFile = options?.tlsCertificateKeyFile?.absoluteString {
+            try self.setUTF8Option(MONGOC_URI_TLSCERTIFICATEKEYFILE, to: certFile)
+        }
+
+        if let password = options?.tlsCertificateKeyFilePassword {
+            try self.setUTF8Option(MONGOC_URI_TLSCERTIFICATEKEYFILEPASSWORD, to: password, redactInErrorMsg: true)
+        }
+    }
+
+    /// Sets and validates authentication-related on the underlying `mongoc_uri_t`.
+    private func applyAndValidateAuthOptions(_ options: MongoClientOptions?) throws {
```

Comment:
I just left these using the specific setters since they were already used in the original implementation. 

that said, for whatever reason, libmongoc doesn't seem to support setting the username and password in that way. it is possible to set the auth source and mechanism that way though. the mechanism properties are a `bson_t`, which there is no generic setter for, so we have to use the property-specific setter for that one too.